### PR TITLE
Redis cache support

### LINKED ISSUE #9280: Add Redis support for Spring cache abstraction
##### **Overview of the feature request**

I have a strong incentive of adding Redis support to JHipster (we have a Prod project on PCF which absolutely need it in the next weeks), however, would you prefer it in the main gen, module or blueprint?

Side question: I noticed JH doesn't generate `@Cacheable`, `@Cacheput` and so on for CRUD: I may also add support for these annotation (but it's a mid term feature).

##### **Motivation for or Use Case**

An Onprem + SaaS microservice architecture with data sync between them: a Rabbit and Redis combination could be a way to help in this case.

##### **Related issues or PR**

#7503
#7691
#1149

-   [x] Checking this box is mandatory (this is just to show you read everything)

### PR DESCRIPTION
**Pull request by Shaolans**
-   Please make sure the below checklist is followed for Pull Requests.

-   [x] [Travis tests](https://travis-ci.org/jhipster/generator-jhipster/pull_requests) are green
-   [x] Tests are added where necessary
-   [ ] Documentation is added/updated where necessary
-   [x] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/master/CONTRIBUTING.md) are followed

This is my implementation of redis cache for the generator. Here you can check a example of an application generated with redis cache https://github.com/Shaolans/jhipster-app-redis-sample.
It uses this branch of Jhipster https://github.com/jhipster/jhipster/pull/324.
Fix #9280

### CODE REVIEW

File: generators/server/templates/src/main/java/package/config/CacheConfiguration.java.ejs [hunk 1]
```diff
@@ -706,4 +727,77 @@ public class CacheConfiguration<% if (cacheProvider === 'hazelcast') { %> implem
         return defaultCache;
     }
     <%_ } _%>
+
+    <%_ if (cacheProvider === 'redis') { _%>
+    private final javax.cache.configuration.Configuration<Object, Object> jcacheConfiguration;
+
+        public CacheConfiguration(JHipsterProperties jHipsterProperties) {
```
Comment (hdurix) [inline generators/server/templates/src/main/java/package/config/CacheConfiguration.java.ejs]: wrong indentation from this line

File: generators/server/templates/src/main/java/package/config/CacheConfiguration.java.ejs [hunk 2]
```diff
@@ -706,4 +727,77 @@ public class CacheConfiguration<% if (cacheProvider === 'hazelcast') { %> implem
...
+            MutableConfiguration<Object, Object> jcacheConfig = new MutableConfiguration<>();
+            Config config = new Config();
+            config.useSingleServer()
+                .setAddress(jHipsterProperties.getCache().getRedis().getServer())
+                .setSubscriptionConnectionMinimumIdleSize(1)
```
Comment (hdurix) [inline generators/server/templates/src/main/java/package/config/CacheConfiguration.java.ejs]: As we use all the default values, could we omit all these configuration options?

**Comment by hdurix** [comment]
@Tcharl @jdubois @anthonyrichir @PierreBesson could you have a look at this PR?

**Comment by PierreBesson** [comment]
There is something strange. It looks like the jhipsterProperties doesn't have the property: `redis.enabled` like what was done for memcached. I think we should stay consistent between the 2 implementations.

File: generators/server/templates/src/main/java/package/config/CacheConfiguration.java.ejs (hunk 2 above)
Comment (Shaolans) [reply generators/server/templates/src/main/java/package/config/CacheConfiguration.java.ejs]: Thanks for your review :+1:

**Comment by Shaolans** [comment]
@PierreBesson We currently only supporting enabling/disabling cache on memcached not for others. Memcached is using Spring Cache and it exists NoOpCacheManager in order to disable the caching. However since this implementation is using JCache I am not sure if we have something similar.

**Comment by DanielFran** [comment]
Normally it is just necessary to add to app.yml:
````txt
spring:
  cache:
    type: none
````

**Comment by hdurix** [comment]
So nothing to do in our hand? Except maybe to explain it in the README?

**Comment by Shaolans** [comment]
@DanielFran Nice thanks !

@hdurix I think so, here we just need to avoid to instantiate a RedisClient when the cache is none but if the cache is disabled, the user should also manually disable hibernate 2nd cache level if there is one since hibernate reuse the same cache.

**Comment by DanielFran** [comment]
@Shaolans I do not believe you need to add your last commit since the cache will not be used if deactivated in configuration file.
I believe we just need to document it and make reference to spring documentation.

**Comment by Shaolans** [comment]
@DanielFran Even if we disable the cache, I think it will still instantiate a redis client and the HibernatePropertiesCustomizer bean will error since there is no cacheManager. Or is it up to the dev to do it ?

**Comment by Shaolans** [comment]
@JackyC415 Hello you need to build the jhipster server library first by running ```./mvnw clean install -Dgpg.skip=true -DskipTests``` on this branch https://github.com/jhipster/jhipster/pull/324

**Comment by pascalgrimaud** [comment]
@Shaolans : just tested it and it works well. Good job ! But I have a comment:
- it looks like I need to start a Redis server before lauching tests, otherwise, all tests will fail
- is it the case ? Can't we use an embedded Redis server or maybe use testcontainers ?

**Comment by pascalgrimaud** [comment]
Another comments too:
- the docs at jhipster.tech need to be updated too
- the option should work for JDL

**Comment by pascalgrimaud** [comment]
If possible, once it's merged, if you can add a dedicated build at https://github.com/hipster-labs/jhipster-daily-builds it would be great !

**Comment by Shaolans** [comment]
> it looks like I need to start a Redis server before lauching tests, otherwise, all tests will fail
> is it the case ? Can't we use an embedded Redis server or maybe use testcontainers ?

You are totally right !

> the docs at jhipster.tech need to be updated too

I was planning on doing it, if the PR was good enough.

> the option should work for JDL

Thanks for reminding me it, I totally forgot that case.

> If possible, once it's merged, if you can add a dedicated build at https://github.com/hipster-labs/jhipster-daily-builds it would be great !

I was planning do it, once merge :) !

I will probably work on these points next week.

**Comment by pascalgrimaud** [comment]
@Shaolans : Thanks. What I prefer in order:
- I don't look how memcached works, it would be nice to be consistent
- embedded server
- disable cache for tests
- testcontainer

**Comment by DanielFran** [comment]
@jdubois Do you have time to review this PR since you were interested in?

**Comment by TonyLuo** [comment]
any update on this? seems still not merged into master branch yet...

**Comment by pascalgrimaud** [comment]
@TonyLuo : there is still an issue with integration tests, that's why it's not ready to be merged yet

File: generators/server/templates/src/main/java/package/config/CacheConfiguration.java.ejs [hunk 3]
```diff
@@ -753,18 +753,17 @@ public class CacheConfiguration<% if (cacheProvider === 'hazelcast') { %> implem
     <%_ } _%>
 
     <%_ if (cacheProvider === 'redis') { _%>
-    private final javax.cache.configuration.Configuration<Object, Object> jcacheConfiguration;
-
-    public CacheConfiguration(JHipsterProperties jHipsterProperties) {
+    @Bean
+    @DependsOn({"testContainerRedis"})
```
Comment (atomfrede) [inline generators/server/templates/src/main/java/package/config/CacheConfiguration.java.ejs]: I don't understand why the production configuration should depend on the testContainer configuration or do I miss anything?

File: generators/server/templates/src/main/java/package/config/CacheConfiguration.java.ejs [hunk 4]
```diff
@@ -773,31 +772,37 @@ public class CacheConfiguration<% if (cacheProvider === 'hazelcast') { %> implem
...
+    private void createCache(javax.cache.CacheManager cm, String cacheName, javax.cache.configuration.Configuration<Object, Object> jcacheConfiguration) {
         javax.cache.Cache<Object, Object> cache = cm.getCache(cacheName);
         if (cache != null) {
             cm.destroyCache(cacheName);
         }
         cm.createCache(cacheName, jcacheConfiguration);
     }
+
+    @Profile("!default")
+    @Bean(name = "testContainerRedis")
```
Comment (atomfrede) [inline generators/server/templates/src/main/java/package/config/CacheConfiguration.java.ejs]: I don't get why there is the need to overwrite the testcontainer bean for production code.

**Comment by atomfrede** [comment]
Good work @Shaolans really looking forward to have it merged!

File: generators/server/templates/src/main/java/package/config/CacheConfiguration.java.ejs (hunk 3 above)
Comment (Shaolans) [reply generators/server/templates/src/main/java/package/config/CacheConfiguration.java.ejs]: Yes you are totally right ! I am still working on it, I need to rework on this part

**Comment by pascalgrimaud** [comment]
Very good job @Shaolans !

**Comment by Shaolans** [comment]
Bug bounty claimed https://opencollective.com/generator-jhipster/expenses/10211

**Comment by sinaghazi** [comment]
for some reason on version v6.10.4 I faced an issue with ExtendWith annotation not being imported in the AccountResourceIT class. My config is Okta and i used Redis for cache.

**Comment by pascalgrimaud** [comment]
@sinaghazi : plz open a new ticket with all information so we can try to reproduce

**Pull request merged**